go-controller/pkg/metrics/master.go

```diff
@@ -232,147 +245,149 @@ var metricPortBindingUpLatency = prometheus.NewHistogram(prometheus.HistogramOpt
-			}, func() float64 {
-				var nbGlobal *nbdb.NBGlobal
-				var probeInterval string
-				var ok bool
-				var err error
-
-				if nbGlobal, err = libovsdbops.FindNBGlobal(nbClient); err != nil {
-					klog.Errorf("Failed to get NB_Global table "+
-						"err: %v", err)
-					return 0
-				}
 
-				if probeInterval, ok = nbGlobal.Options["northd_probe_interval"]; !ok {
-					klog.Errorf("Failed to get northd_probe_interval from NB_Global table options column")
-					return 0
-				}
+			return parseMetricToFloat(MetricOvnSubsystemNorthd, "probe_interval", probeInterval)
+		},
+	))
+}
 
-				return parseMetricToFloat(MetricOvnSubsystemNorthd, "probe_interval", probeInterval)
-			},
-		))
-	})
+// RegisterMasterFunctional is a collection of metrics that help us understand ovnkube-master functions. Call once after
+// LE is won.
+func RegisterMasterFunctional() {
+	// No need to unregister because process exits when leadership is lost.
+	prometheus.MustRegister(metricV4HostSubnetCount)
+	prometheus.MustRegister(metricV6HostSubnetCount)
+	prometheus.MustRegister(metricV4AllocatedHostSubnetCount)
+	prometheus.MustRegister(metricV6AllocatedHostSubnetCount)
+	prometheus.MustRegister(metricEgressIPCount)
+	prometheus.MustRegister(metricEgressFirewallRuleCount)
+	prometheus.MustRegister(metricEgressFirewallCount)
+	prometheus.MustRegister(metricEgressRoutingViaHost)
 }
 
-// StartMasterMetricUpdater adds a goroutine that updates a "timestamp" value in the
-// nbdb every 30 seconds. This is so we can determine freshness of the database.
-// Also, update IPsec enabled or disable metric.
-func StartMasterMetricUpdater(stopChan <-chan struct{}, nbClient libovsdbclient.Client) {
-	startMasterMetricUpdaterOnce.Do(func() {
-		addIPSecMetricHandler(nbClient)
-		go func() {
-			ticker := time.NewTicker(30 * time.Second)
-			defer ticker.Stop()
-			for {
-				select {
-				case <-ticker.C:
-					updateE2ETimestampMetric(nbClient)
-				case <-stopChan:
-					return
+// RunTimestamp adds a goroutine that registers and updates timestamp metrics.
+// This is so we can determine 'freshness' of the components NB/SB DB and northd.
+// Function must be called once.
+func RunTimestamp(stopChan <-chan struct{}, sbClient, nbClient libovsdbclient.Client) {
+	// Metric named nb_e2e_timestamp is the UNIX timestamp this instance wrote to NB DB. Updated every 30s with the
+	// current timestamp.
+	prometheus.MustRegister(metricNbE2eTimestamp)
+
+	// Metric named sb_e2e_timestamp is the UNIX timestamp observed in SB DB. The value is read from the SB DB
+	// cache when metrics HTTP endpoint is scraped.
+	scrapeOvnSbE2eTimestamp := func() float64 {
+		sbGlobal, err := libovsdbops.FindSBGlobal(sbClient)
+		if err != nil {
+			klog.Errorf("Failed to get global options for the SB_Global table: %v", err)
+			return 0
+		}
+		if val, ok := sbGlobal.Options["e2e_timestamp"]; ok {
+			return parseMetricToFloat(MetricOvnkubeSubsystemMaster, "sb_e2e_timestamp", val)
+		}
+		return 0
```

Comment:
How can't we use here the `getDbOptionsTimestamp` function or somehow share a bit of code between the two? They are very similar.

Reply:
Its duplication of code. Fixed.